src/main/scala/is/hail/io/plink/ExportPlink.scala

```diff
@@ -1,95 +1,183 @@
+      val hConf = sHConfBc.value.value
+      val f = partFile(d, i, TaskContext.get)
+      val bedPartPath = tmpBedDir + "/" + f
+      val bimPartPath = tmpBimDir + "/" + f
+      var rowCount = 0L
 
-    val famRows = vsm
-      .colValues.value
-      .map { sa =>
-        ec.set(0, sa)
-        val a = f().map(Option(_))
-        famFns.map(_ (a)).mkString("\t")
+      hConf.writeTextFile(bimPartPath) { bimOS =>
+        hConf.writeFile(bedPartPath) { bedOS =>
+          val v = new RegionValueVariant(fullRowType)
+          val a = new BimAnnotationView(fullRowType)
+          val hcv = HardCallView(fullRowType)
+          val bp = new BitPacker(2, (i: Int) => bedOS.write(i))
+
+          it.foreach { rv =>
+            v.setRegion(rv)
+            a.setRegion(rv)
+            ExportPlink.writeBimRow(v, a, bimOS)
+
+            hcv.setRegion(rv)
+            ExportPlink.writeBedRow(hcv, bp, nSamples)
+
+            rowCount += 1
+          }
+        }
       }
 
-    vsm.hc.hadoopConf.writeTextFile(path + ".fam")(out =>
-      famRows.foreach(line => {
-        out.write(line)
-        out.write("\n")
-      }))
+      Iterator.single(rowCount)
+    }.collect().sum
+
+    hConf.writeFile(tmpBedDir + "/_SUCCESS")(out => ())
+    hConf.writeFile(tmpBedDir + "/header")(out => out.write(ExportPlink.bedHeader))
+    hConf.copyMerge(tmpBedDir, path + ".bed", nPartitions, header = true)
+
+    hConf.writeTextFile(tmpBimDir + "/_SUCCESS")(out => ())
+    hConf.copyMerge(tmpBimDir, path + ".bim", nPartitions, header = false)
+
+    mv.colsTableValue.export(path + ".fam", header = false)
+
+    info(s"wrote $nRecordsWritten variants and $nSamples samples to '$path'")
+  }
+}
+
+class BimAnnotationView(rowType: TStruct) extends View {
+  private val varidField = rowType.fieldByName("varid")
+  private val posMorganField = rowType.fieldByName("pos_morgan")
+  private val varidIdx = varidField.index
+  private val posMorganIdx = posMorganField.index
+  private val tvarid = varidField.typ.asInstanceOf[TString]
+  private val tposm = posMorganField.typ.asInstanceOf[TInt32]
+  private var region: Region = _
+  private var varidOffset: Long = _
+  private var posMorganOffset: Long = _
+
+  private var cachedVarid: String = _
+
+  def setRegion(region: Region, offset: Long) {
+    this.region = region
+
+    assert(rowType.isFieldDefined(region, offset, varidIdx))
+    assert(rowType.isFieldDefined(region, offset, posMorganIdx))
+    this.varidOffset = rowType.loadField(region, offset, varidIdx)
+    this.posMorganOffset = rowType.loadField(region, offset, posMorganIdx)
+
+    cachedVarid = null
+  }
+
+  def positionMorgan(): Int =
+    region.loadInt(posMorganOffset)
+
+  def varid(): String = {
+    if (cachedVarid == null)
+      cachedVarid = TString.loadString(region, varidOffset)
+    cachedVarid
+  }
+}
+
+class BitPacker(nBitsPerItem: Int, consume: (Int) => Unit) extends Serializable {
```

Comment:
This is my last comment, sorry for spreading things out and making it difficult for you! I had trouble understanding what this did until I found the usage. I think it makes a bit more sense to pass the output stream as an argument rather than a function. Probably way faster than a virtual function call here too!

Reply:
I had wanted to make this class usable for other consumers (such as ArrayBuilder), but I think for now your suggestion is the correct thing to do. If we ever revisit export_bgen, we'll probably want to modify this class so it can be used there too.